src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/ReadOnlyListExtensions.cs

```diff
@@ -36,6 +52,49 @@ static ImmutableArray<TResult> BuildResult(IReadOnlyList<T> items, Func<T, TResu
         }
     }
 
+    /// <summary>
+    ///  Projects each element of an <see cref="IReadOnlyList{T}"/> into a new form by incorporating the element's index.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements in <paramref name="list"/>.</typeparam>
+    /// <typeparam name="TResult">The type of the value returned by <paramref name="selector"/>.</typeparam>
+    /// <param name="list">An <see cref="IReadOnlyList{T}"/> of values to invoke a transform function on.</param>
+    /// <param name="selector">
+    ///  A transform function to apply to each element; the second parameter of the function represents the index of the element.
+    /// </param>
+    /// <returns>
+    ///  Returns a new <see cref="ImmutableArray{T}"/> whose elements are the result of invoking the transform function
+    ///  on each element of <paramref name="list"/>.
+    /// </returns>
+    public static ImmutableArray<TResult> SelectAsArray<T, TResult>(this IReadOnlyList<T> list, Func<T, int, TResult> selector)
+    {
+        if (list is ImmutableArray<T> array)
+        {
+            return ImmutableArrayExtensions.SelectAsArray(array, selector);
+        }
+
+        return list switch
+        {
+            [] => [],
+            [var item] => [selector(item, 0)],
+            [var item1, var item2] => [selector(item1, 0), selector(item2, 1)],
+            [var item1, var item2, var item3] => [selector(item1, 0), selector(item2, 1), selector(item3, 2)],
+            [var item1, var item2, var item3, var item4] => [selector(item1, 0), selector(item2, 1), selector(item3, 2), selector(item4, 3)],
+            var items => BuildResult(items, selector)
+        };
+
+        static ImmutableArray<TResult> BuildResult(IReadOnlyList<T> items, Func<T, int, TResult> selector)
+        {
+            using var results = new PooledArrayBuilder<TResult>(capacity: items.Count);
```

Comment:
>using var results = new PooledArrayBuilder<TResult>(capacity: items.Count); [](http://example.com/codeflow?start=12&length=75)

consider using new/ImmutableCollectionsMarshal in the SelectAsArray methods

Reply:
You're absolutely right. Since we have an `IReadOnlyList<T>`, we can just create the result array up front and not do the extra work of checking the number of items.